contribute/design/1.x/cstor/pdb/20200113-cstor-poddisruptionbudget.md

```diff
@@ -0,0 +1,170 @@
+
+## Proposal
+
+### User Stories
+
+#### HighAvailable Volumes Shouldn't be disturbed for Cluster Upgrade Activities
+As an OpenEBS user, HA volumes shouldn't be disturbed for cluster upgrade or maintenance
+activities.
+
+### Proposed Implementation
+
+Kubernetes will send volume creation request to CSI driver when PVC is created.
+CSI driver will send a cStorVolume provision request via CVC CR to cStorVolumeConfig(CVC)
+controller. CVC controller will schedule the replicas among the pools after that
+CVC controller will create a PDB among the pool pods where the volume replicas
+are scheduled.
+
+Example:
+- Provisioned a CSPC with five pool specs that intern creates 5 cStor Pools.
+- Now, when 'n' no.of HA volumes are provisioned(with replica count as 3) then
+  volume replicas can be schedule in any available cStor pools.
+- After volume replicas are placed then cStorVolumeConfig controller will
+  create PDB among those cStor pools.
+- After successful creation of PDB corresponding cStorVolumeConfig will be
+  updated with PDB label.
+
+So when `n` no.of HA volumes are provisioned then sigma(nCr)[where n is no.of pools,
+r is no.of replicas] no.of PDB's will be created. For example if `n` pools are
+created then at max nC3 + nC4 + nC5 PDB's will be created(Where 3, 4 & 5 are
+replicas).
+
+Note: PDB will be created only for HA volumes(i.e only for the volumes greater
+than or equal to 3 replica count).
+
+### Steps to perform user stories
+- User has to create HighAvailable volumes on cStor then PDB will be created by
+  default.
+
+### Low Level Design
+
+#### Work Flow
+- User/Admin provisioned HA volume, Now CVC controller schedules the replicas in
+  the cStor pools. After scheduling the replicas CVC controller will try to get
+  PDB that was created among those cStor pools. If PDB doesn't exist then PDB will
+  be created with following details:
+    - All the pool names where the volume replicas are scheduled will be added as
+      labels on PDB[Why? to identify the PDB without iterating over all the existing PDBS].
+    - PDB name will be added as label(openebs.io/pod-disruption-budget: <pdb_name>)
+      on CVC name will be added as annotation[why? To identify how many volumes
+      are referring particular PDB. So during deprovisioning time if there are no
+      reference to PDB then PDB can be deleted].
+- During deprovision of volume CVC controller will verify is there any other
+  volume referring to PDB created among those pools. If such volume exists then
+  PDB will be destroyed and then finalizers will be removed from CVC or else
+  finalizers will be removed from CVC.
+
+#### Sample Yaml
+Example PDB yaml:
+```yaml
+apiVersion: policy/v1beta1
+kind: PodDisruptionBudget
+metadata:
+  name: cspc-name-<hash>
+  labels:
+    openebs.io/<cstor-pool1>:
+    openebs.io/<cstor-pool2>:
+    openebs.io/<cstor-pool3>:
+spec:
+  minAvailable: replicaCount/2 + 1  ###[We can also have maxUnavailable as 1].
```

Comment:
maxUnavailable as `replicaCount/2 - 1`

Reply:
it would be `upperCeil(replicaCount/2) - 1`

Reply:
I think to optimize more instead of creating unnecessary resources then it seems like we need to populate `maxUnavailable as 1`.